proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,322 @@
+>
+> ```kotlin
+> class MyClass {
+>     private val _data = ArrayList<String>()
+>     val data: List<String>
+>         get() = _data
+> }
+> ```
+
+With the proposed syntax in mind, the above code snippet could be rewritten as follows:
+
+```kotlin
+class MyClass {
+    val data: List<String>
+        field = ArrayList()
+}
+```
+
+## Use Cases
+
+### Motivation
+
+> - There is no clean idiomatic way to do this pattern in Kotlin
+> - Current best approach throws away all benefits of Kotlin properties and forces developer to write Java-like code with separate private field and public getter
+> - Current best approach forces developer to assign two different names to single property (or pad private property, for example adding `_` prefix and then using this prefix everywhere in code)
+> - This is another place where Java pattern could be introduced into Kotlin with less boilerplate
+
+### Read-Only from Outside
+
+> We often do not want our data structures to be modified from outside. Unlike Java, this can be easily achieved in Kotlin by just exposing read-only `List`. But as already ilustrated in above example, exposing different type of a property is a bit messy.
+
+```kotlin
+private val _items = mutableListOf<Item>()
+val item : List<Item> by _items
+```
+
+And the new syntax allows us to write:
+
+```kotlin
+val items: List<Item>
+	private field = mutableListOf()
+```
+
+### Android Architecture Components
+
+>  Proper way to do Architecture components is to use `MutableLiveData` (`LiveData` implementation that allows caller to change its value) privately inside View Model classes and then only expose read-only `LiveData` objects outside.
+
+Sample code [from an Android app](https://github.com/elpassion/crweather/blob/9c3e3cb803b7e4fffbb010ff085ac56645c9774d/app/src/main/java/com/elpassion/crweather/MainModel.kt#L14):
+
+```kotlin
+private val mutableCity = MutableLiveData<String>().apply { value = "" }
+private val mutableCharts = MutableLiveData<List<Chart>>().apply { value = emptyList() }
+private val mutableLoading = MutableLiveData<Boolean>().apply { value = false }
+private val mutableMessage = MutableLiveData<String>()
+
+val city: LiveData<String> = mutableCity
+val charts: LiveData<List<Chart>> = mutableCharts
+val loading: LiveData<Boolean> = mutableLoading
+val message: LiveData<String> = mutableMessage
+```
+
+Becomes:
+
+```kotlin
+val city: LiveData<String>
+	field = MutableLiveData().apply { value = "" }
+val charts: LiveData<List<Chart>>
+	field = MutableLiveData().apply { value = emptyList() }
+val loading: LiveData<Boolean>
+	field = MutableLiveData().apply { value = false }
+val message: LiveData<String>
+	field = MutableLiveData()
+```
+
+### RX Observable and Subjects
```

Comment:
Since it is already the year 2021 (and also taking into account readers from the future), it is way better to give this example using Kotlin Flow instead of Rx, e.g. with `MutableStateFlow<String>` that is exposed as `StateFlow<String>`.

Reply:
This example would, indeed, illustrate the use of an explicit backing field, but since it's typical to use `asStateFlow()`, the smart type narrowing wouldn't work, and the example would be a bit misleading (apart from the STN there's currently no way to access the field value from outside the accessors)